sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/ForbidCertainMethodsCheck.java

```diff
@@ -0,0 +1,176 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.github.sevntu.checkstyle.checks.coding;
+
+import java.util.regex.Pattern;
+
+import com.github.sevntu.checkstyle.Utils;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * Check that forbidden methods are not used. We can forbid methods by ruleName and number of
+ * arguments.
+ * This can be used to enforce things like:
+ * <ul>
+ * <li> System.exit() should not be called.</li>
+ * <li> Assert.assertTrue() and Assert.assertFalse() have a 1 arg variant that does not provide a
+ * helpful message on failure.
+ * </ul>
+ * Parameters are:
+ * <ul>
+ * <li><b>methodName</b> - This is name of the method to be forbidden. It can be a regex.
+ * When blank or unspecified, all the methods are allowed.</li>
+ * <li><b>argumentCount</b> - Number of arguments the method takes. It can be a regex.
```

Comment:
> Number .... regexp 

It is really weird. It is a regexp, no doubts.
please explain a reason of RegExp for Number. Please provide example to user. As it is far from obvious.

Reply:
Consider the following series of overloaded methods.
```java
public static <T> T execute(Callable<T> task, int tries, int waitTime, int timeout, TimeUnit waitTimeUnit)
public static <T> T execute(Callable<T> task, int tries, int waitTimeSec, int timeoutSec)
public static <T> T execute(Callable<T> task, int tries, int waitTimeSec)
public static <T> T execute(Callable<T> task)
```
Last three methods execute the first method with some default value of wait time, timeout etc at level of second. For my application and we want retry at millisecond level which means that only first method is useful. So, I want to forbid other three methods throughout my codebase. Having argumentCount to be regex will make this easy.

Reply:
Please provide example to javadoc on how user can benefit from RegExp here ... I still do not see how you going to use it.
Please place some other examples here PR for me to review.

Reply:
@romani For the above you can set argumentCount to `1|3|4` or `[^5]`
The above example is inspired by a closed source project, and it's javadoc is not published on internet.

Reply:
This is not good reason to use RegExp. It is better to ask user have to use several instances of module for each `argumentCount`. RegExp for numbers is weird stuff. As "1|3|4" is matching to 11, 34, 43 - making any `^5$` - is very weird value for `argumentCount`.

@rnveach , do you have have any comments on this ?

Reply:
@romani I don't agree with several instances. Like I stated in main issue, user could want to print violation on vararg method. There is no amount of instances where you can make to forbid vararg without some type of range or regular expression as variable arguments can receive an infinite amount of arguments.

Instead of regular expression, how about we implement some type of multi-range value?
Example: `1` or `2-3` or `4+` or `1-3,5+` or etc...

> "1|3|4" is matching to 11, 34, 43 - making any ^5$ - is very weird value for argumentCount.

Expresssion should be a full match if we leave it as is and not require `^$`.

Reply:
>Instead of regular expression, how about we implement some type of multi-range value?

Exactly this target I waited to be raised. Example of infinite arguments has to be in javadoc.

>Example: 1 or 2-3 or 4- (meaning 4 to infinity) or 1-3,5- (meaning only allow 0 or 4 arguments)

I more in favor of this format, rather than regexp. To keep user ability to configure on leash.

@rnveach , @raghavgautam , does it work for you ?

Reply:
@romani Since I brought it up, it works for me.

Reply:
@romani @rnveach 
> Example of infinite arguments has to be in javadoc.

The exit method example that I had put in javadoc applies to infinite arguments.
For range formating, the benefit of regex is that it's familiar and well known but a bit awkward for larger numbers. The benefit of multi-range format is that it readable. So, I don't have a strong preference either way.
@romani @rnveach Is there any other check that is deals with a ranges of numbers ?

Reply:
> Is there any other check that deals with a ranges of numbers ?

To my knowledge, no. The closest is https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/filters/IntRangeFilter.java used by SuppressElement. But it doesn't support open-ended bounds.

> but a bit awkward for larger numbers

It is also akward for numbers that have different sized digits as you can't do `[5-16]` but have to define it as `[5-9]|1[0-6]` which is what made decide on the range.